### PR TITLE
feat: Adding --cluster-cancel and --cluster-cancel-nargs

### PR DESCRIPTION
**Pull request by holtgrewe**
### Description

This PR adds the optional arguments `--cluster-cancel` and `--cluster-mcancel`. The user can specify commands for terminating jobs by their job ID, either one at a time (`--cluster-cancel`) or multiple at once (`--cluster-mcancel`). Testing this was somewhat tricky because you have to simulate senting `SIGINT` to the job so I ended extending `tests/common.py` a bit.

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of 

Addresses: #570

### CODE REVIEW

File: snakemake/executors/__init__.py [hunk 1]
```diff
@@ -976,8 +980,38 @@ def __init__(
...
+                chunks = list(_chunks(jobids, max_mcancel))
+            else:
+                cmd = self.cancelcmd
+                chunks = list(_chunks(jobids, 1))
+            # Go through the chunks and cancel the jobs.
+            for chunk in chunks:
+                try:
+                    subprocess.Popen([cmd] + chunk, shell=False)
+                except:
+                    # This is common - logging a warning would probably confuse the user.
```
Comment (johanneskoester) [inline snakemake/executors/__init__.py:1007]: True, but we should collect the number of failed cancel commands, and issue a warning at the end that tells the user how many of them failed (if that number is greater than 0).

File: tests/common.py [hunk 2]
```diff
@@ -156,16 +159,24 @@ def run(
...
-            )
-            success = True
+            if sigint_after is None:
+                subprocess.check_output(
+                    shellcmd, cwd=path if no_tmpdir else tmpdir, shell=True
+                )
+                success = True
+            else:
+                with subprocess.Popen(
+                    shellcmd.split(" "), cwd=path if no_tmpdir else tmpdir
```
Comment (johanneskoester) [inline tests/common.py:169]: I think the idiomatic way is to use `shutil.split`.

**Review by johanneskoester — CHANGES_REQUESTED** [review]
Thanks a lot, very nice work! The Snakefile of your testcase is missing. Apart from that, a few requests below.

File: snakemake/executors/__init__.py (hunk 1 above)
Comment (holtgrewe) [reply snakemake/executors/__init__.py:1007]: Agreed, this was here before but will be fixed.

File: snakemake/executors/__init__.py [hunk 3]
```diff
@@ -976,8 +980,38 @@ def __init__(
             )
 
     def cancel(self):
-        logger.info("Will exit after finishing currently running jobs.")
-        self.shutdown()
+        # maximum number of jobs to cancel at once
+        max_mcancel = 1000
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply snakemake/executors/__init__.py:984]: Good point, will add.

**Comment by holtgrewe** [comment]
Sorry for the many pushes but I had a hard time uncovering why the CI was failing and it was running locally (as so often, timing issues because of fewer cores in CI than in dev). I'm positive that this is ready for re-review :eyeglasses: :mag: :microscope:

**Comment by holtgrewe** [comment]
:tada:

**Review by johanneskoester — APPROVED** [review]
Excellent, awesome and clean work, as always!

**Pull request merged**